dm/simulator/internal/mcp/mcp.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mcp defines the Modification Candidate Pool (MCP).
+package mcp
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+)
+
+// ModificationCandidatePool is the core container storing all the current unique keys for a table.
+type ModificationCandidatePool struct {
+	sync.RWMutex
+	keyPool  []*UniqueKey
+	theRand  *rand.Rand
+	randLock sync.Mutex
+}
+
+// NewModificationCandidatePool create a new MCP.
+func NewModificationCandidatePool(capcity int) *ModificationCandidatePool {
+	theKeyPool := make([]*UniqueKey, 0, capcity)
+	theRand := rand.New(rand.NewSource(time.Now().Unix()))
+	return &ModificationCandidatePool{
+		keyPool: theKeyPool,
+		theRand: theRand,
+	}
+}
+
+// NextUK randomly picks a unique key in the MCP.
+func (mcp *ModificationCandidatePool) NextUK() *UniqueKey {
+	mcp.RLock()
+	defer mcp.RUnlock()
+	if len(mcp.keyPool) == 0 {
+		return nil
+	}
+	mcp.randLock.Lock()
+	idx := mcp.theRand.Intn(len(mcp.keyPool))
+	mcp.randLock.Unlock()
+	return mcp.keyPool[idx] // pass by reference
+}
+
+// Len gets the current length of the MCP.
+func (mcp *ModificationCandidatePool) Len() int {
+	mcp.RLock()
+	defer mcp.RUnlock()
+	return len(mcp.keyPool)
+}
+
+// AddUK adds the unique key into the MCP.
+// It has side effect: the input UK's row ID will be changed.
+func (mcp *ModificationCandidatePool) AddUK(uk *UniqueKey) error {
+	mcp.Lock()
+	defer mcp.Unlock()
+	if len(mcp.keyPool)+1 > cap(mcp.keyPool) {
```

Comment:
This is to limit the overall size of the upstream simulating tables so that the storage is not overwhelmed for some simulating workloads like continuous INSERT operations.  In the design, if the MCP is touching the capacity threshold, an eviction process will be triggered.  Currently the eviction logic is not implemented.  So I just report an error here.